java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -491,11 +492,46 @@ public void visitLambdaExpression(LambdaExpressionTree tree) {
         }
       }
       super.visitLambdaExpression(tree);
+      if(lambdaType.isUnknown() || lambdaType.isTagged(JavaType.DEFERRED)) {
+        return;
+      }
+      refineLambdaType(lambdaExpressionTree, lambdaType);
     } else {
       registerType(tree, symbols.deferedType(lambdaExpressionTree));
     }
   }
 
+  private void refineLambdaType(LambdaExpressionTreeImpl lambdaExpressionTree, JavaType lambdaType) {
+    JavaType samReturnType = (JavaType) getSamMethod(lambdaType).returnType().type();
+    JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, lambdaType);
```

Comment:
To get the captured return type, I thunk it's more correct to proceed like this:
```
    JavaType capturedReturnType = null;
    if (lambdaType.isParameterized() && samReturnType.isTagged(JavaType.TYPEVAR)) {
      capturedReturnType = ((ParametrizedTypeJavaType) lambdaType).substitution((TypeVariableJavaType) samReturnType);
    }
```